### ISSUE TITLE
perl ignores errors when asked to invoke or syntax check a directory

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9597]
Migrated from [rt.perl.org#61362](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=61362) (status was 'resolved')

Searchable as RT61362$

**Comment by p5pRT** [comment]
### From perlbugcrowell08@lightandmatter.com

Hi\,

I'd like to report a bug in perl\, which is that if foo is a directory\,
"perl \-c foo" always responds with "foo syntax OK\." \(The most common
way this comes up is if you have a program foo that lives in a directory
called foo\, so you intended "perl \-c foo/foo\," but accidentally did
"perl \-c foo" instead\.\)

&nbsp;    Regards\,

&nbsp;    Ben



---
output of perl \-V


---
Summary of my perl5 \(revision 5 version 10 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.6\.24\-15\-server\, archname=x86\_64\-linux\-gnu\-thread\-multi
&nbsp;    uname='linux crested 2\.6\.24\-15\-server \#1 smp mon apr 7 17&#8203;:10&#8203;:32 utc 2008 x86\_64 gnulinux '
&nbsp;    config\_args='\-Dusethreads \-Duselargefiles \-Dccflags=\-DDEBIAN \-Dcccdlflags=\-fPIC \-Darchname=x86\_64\-linux\-gnu \-Dprefix=/usr \-Dprivlib=/usr/share/perl/5\.10 \-Darchlib=/usr/lib/perl/5\.10 \-Dvendorprefix=/usr \-Dvendorlib=/usr/share/perl5 \-Dvendorarch=/usr/lib/perl5 \-Dsiteprefix=/usr/local \-Dsitelib=/usr/local/share/perl/5\.10\.0 \-Dsitearch=/usr/local/lib/perl/5\.10\.0 \-Dman1dir=/usr/share/man/man1 \-Dman3dir=/usr/share/man/man3 \-Dsiteman1dir=/usr/local/man/man1 \-Dsiteman3dir=/usr/local/man/man3 \-Dman1ext=1 \-Dman3ext=3perl \-Dpager=/usr/bin/sensible\-pager \-Uafs \-Ud\_csh \-Ud\_ualarm \-Uusesfio \-Uusenm \-DDEBUGGING=\-g \-Doptimize=\-O2 \-Duseshrplib \-Dlibperl=libperl\.so\.5\.10\.0 \-Dd\_dosuid \-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fno\-strict\-aliasing \-pipe \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2 \-g'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fno\-strict\-aliasing \-pipe \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.3\.1'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
&nbsp;    libs=\-lgdbm \-lgdbm\_compat \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    perllibs=\-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    libc=/lib/libc\-2\.8\.90\.so\, so=so\, useshrplib=true\, libperl=libperl\.so\.5\.10\.0
&nbsp;    gnulibc\_version='2\.8\.90'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2 \-g \-L/usr/local/lib'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: MULTIPLICITY PERL\_DONT\_CREATE\_GVSV
&nbsp;                        PERL\_IMPLICIT\_CONTEXT PERL\_MALLOC\_WRAP USE\_64\_BIT\_ALL
&nbsp;                        USE\_64\_BIT\_INT USE\_ITHREADS USE\_LARGE\_FILES
&nbsp;                        USE\_PERLIO USE\_REENTRANT\_API
&nbsp;  Built under linux
&nbsp;  Compiled at Jul 24 2008 09&#8203;:03&#8203;:56
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /etc/perl
&nbsp;    /usr/local/lib/perl/5\.10\.0
&nbsp;    /usr/local/share/perl/5\.10\.0
&nbsp;    /usr/lib/perl5
&nbsp;    /usr/share/perl5
&nbsp;    /usr/lib/perl/5\.10
&nbsp;    /usr/share/perl/5\.10
&nbsp;    /usr/local/lib/site\_perl
&nbsp;    \.

**Comment by p5pRT** [comment]
### From @ikegami

On Sun\, Dec 14\, 2008 at 2&#8203;:30 PM\, via RT Ben Crowell \<
perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> \# New Ticket Created by  Ben Crowell
> \# Please include the string&#8203;:  \[perl \#61362\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=61362 >
>
>
> Hi\,
>
> I'd like to report a bug in perl\, which is that if foo is a directory\,
> "perl \-c foo" always responds with "foo syntax OK\." \(The most common
> way this comes up is if you have a program foo that lives in a directory
> called foo\, so you intended "perl \-c foo/foo\," but accidentally did
> "perl \-c foo" instead\.\)
>
>        Regards\,
>
>        Ben
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> output of perl \-V
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> Summary of my perl5 \(revision 5 version 10 subversion 0\) configuration&#8203;:
>  Platform&#8203;:
>    osname=linux\, osvers=2\.6\.24\-15\-server\,
> archname=x86\_64\-linux\-gnu\-thread\-multi


FYI\, it could be one of ActivePerl's patches\, but Windows actually gets it
right\. Tested with both 5\.8\.8 and 5\.10\.0\. \(perl \-v below\)

>perl \-c foo
Can't open perl script "foo"&#8203;: Permission denied


This is perl\, v5\.8\.8 built for MSWin32\-x86\-multi\-thread
\(with 12 registered patches\, see perl \-V for more detail\)

Copyright 1987\-2007\, Larry Wall

Binary build 824 \[287188\] provided by ActiveState http&#8203;://www\.ActiveState\.com
Built Sep  3 2008 11&#8203;:14&#8203;:55


This is perl\, v5\.10\.0 built for MSWin32\-x86\-multi\-thread
\(with 1 registered patch\, see perl \-V for more detail\)

Copyright 1987\-2007\, Larry Wall

Binary build 1001 \[283495\] provided by ActiveState
http&#8203;://www\.ActiveState\.com
Built Dec 18 2007 08&#8203;:46&#8203;:15

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @iabyn

On Sun\, Dec 14\, 2008 at 09&#8203;:28&#8203;:39PM \-0500\, Eric Brine wrote&#8203;:
> FYI\, it could be one of ActivePerl's patches\, but Windows actually gets it
> right\. Tested with both 5\.8\.8 and 5\.10\.0\. \(perl \-v below\)
> 
> >perl \-c foo
> Can't open perl script "foo"&#8203;: Permission denied

With bleed on linux\, I get silent "success"&#8203;:

$ \./perl \-c /tmp
/tmp syntax OK

An strace shows that it is ignoring a read error while trying to
read the directory&#8203;:

&nbsp;    open\("/tmp"\, O\_RDONLY|O\_LARGEFILE\)      = 3
&nbsp;    \.\.\.
&nbsp;    read\(3\, 0x851973c\, 4096\)                = \-1 EISDIR \(Is a directory\)
&nbsp;    close\(3\)                                = 0
&nbsp;    write\(2\, "/tmp syntax OK\\n"\, 15\)        = 15


\-\- 
My Dad used to say 'always fight fire with fire'\, which is probably why
he got thrown out of the fire brigade\.

**Comment by p5pRT** [comment]
### From @eserte

Dave Mitchell \<davem@&#8203;iabyn\.com> writes&#8203;:

> On Sun\, Dec 14\, 2008 at 09&#8203;:28&#8203;:39PM \-0500\, Eric Brine wrote&#8203;:
>> FYI\, it could be one of ActivePerl's patches\, but Windows actually gets it
>> right\. Tested with both 5\.8\.8 and 5\.10\.0\. \(perl \-v below\)
>> 
>> >perl \-c foo
>> Can't open perl script "foo"&#8203;: Permission denied
>
> With bleed on linux\, I get silent "success"&#8203;:
>
> $ \./perl \-c /tmp
> /tmp syntax OK
>
> An strace shows that it is ignoring a read error while trying to
> read the directory&#8203;:

On FreeBSD\, it is actually possible to \_execute\_ a directory if it has
an appropriate filename\. First create directory and file&#8203;:

mkdir "/tmp/dirhack";
open my $fh\, ">"\, qq\{/tmp/dirhack/\\n\#\!perl\\nprint "Hello\, world\!\\\\n";\\n\_\_END\_\_\\n\}
&nbsp;    or die $\!;
\_\_END\_\_

and then run

&nbsp;     perl \-x /tmp/dirhack

Regards\,
&nbsp;    Slaven

\-\- 
Slaven Rezic \- slaven \<at> rezic \<dot> de
&nbsp;  BBBike \- route planner for cyclists in Berlin
&nbsp;  WWW version&#8203;:                           http&#8203;://www\.bbbike\.de
&nbsp;  Perl/Tk version for Unix and Windows&#8203;:  http&#8203;://bbbike\.sourceforge\.net

**Comment by p5pRT** [comment]
### From @jmdh

This isn't specific to '\-c' so I have updated the subject accordingly\.
This was also reported as 

\<http&#8203;://bugs\.debian\.org/cgi\-bin/bugreport\.cgi?bug=689412>

**Comment by p5pRT** [comment]
### From @rjbs

I have written b1469bd to catch this\.  I haven't written any tests\, but can\.  I have tested it by 
hand\.  I thought it worth letting anybody who knows perl\.c better than I do \(read&#8203;: anybody\) have 
a look before I spend more time on it\.

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat Jan 05 17&#8203;:53&#8203;:13 2013\, rjbs wrote&#8203;:
> I have written b1469bd to catch this\.  I haven't written any tests\,
> but can\.  I have tested it by
> hand\.  I thought it worth letting anybody who knows perl\.c better than
> I do \(read&#8203;: anybody\) have
> a look before I spend more time on it\.

First attempt at a test\, attached\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @jkeenan

<details><summary>0001-Add-a-test-to-detect-error-when-attempting-to-syntax.patch</summary>

```diff
From 975cb0aa2e4cb19f7fe47ba3f9f3caaa2399973a Mon Sep 17 00:00:00 2001
From: James E Keenan <jkeenan@cpan.org>
Date: Sat, 5 Jan 2013 22:25:29 -0500
Subject: [PATCH] Add a test to detect error when attempting to syntax-check a directory.

Not yet working; todo_skip it.

For: RT #61362
---
 t/run/switches.t |   14 +++++++++++++-
 1 files changed, 13 insertions(+), 1 deletions(-)

diff --git a/t/run/switches.t b/t/run/switches.t
index 5c3b5ee..bb46e6a 100644
--- a/t/run/switches.t
+++ b/t/run/switches.t
@@ -11,7 +11,7 @@ BEGIN {
 
 BEGIN { require "./test.pl"; }
 
-plan(tests => 114);
+plan(tests => 115);
 
 use Config;
 
@@ -107,6 +107,18 @@ SWTEST
     );
 }
 
+TODO: {
+    todo_skip("RT #61362: perl ignores errors when asked to invoke or syntax check a directory");
+    my $tempdir = tempfile;
+    mkdir $tempdir, 0700 or die "Can't mkdir '$tempdir': $!";
+    like(
+        runperl( switches => [ '-c' ], args  => [ $tempdir ], stderr => 1),
+        qr/Can't open perl script.*$tempdir.*is a directory/s,
+        "RT \#61362: Cannot syntax-check a directory"
+    );
+    rmdir $tempdir or die "Can't rmdir '$tempdir': $!";
+}
+
 # Tests for -l
 
 $r = runperl(
-- 
1.6.3.2


```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat Jan 05 19&#8203;:29&#8203;:55 2013\, jkeenan wrote&#8203;:
> On Sat Jan 05 17&#8203;:53&#8203;:13 2013\, rjbs wrote&#8203;:
> > I have written b1469bd to catch this\.  I haven't written any tests\,
> > but can\.  I have tested it by
> > hand\.  I thought it worth letting anybody who knows perl\.c better than
> > I do \(read&#8203;: anybody\) have
> > a look before I spend more time on it\.
> 
> First attempt at a test\, attached\.
> 
> Thank you very much\.
> Jim Keenan

Second attempt at a patch\.  I don't think the test needed to be todo\-ed\.

**Comment by p5pRT** [comment]
### From @jkeenan

<details><summary>0001-Add-a-test-to-detect-error-when-attempting-to-syntax.patch</summary>

```diff
From d0f59158e7938543e82e306c56f927866a506ed8 Mon Sep 17 00:00:00 2001
From: James E Keenan <jkeenan@cpan.org>
Date: Sat, 5 Jan 2013 22:25:29 -0500
Subject: [PATCH] Add a test to detect error when attempting to syntax-check a directory.

For: RT #61362
---
 t/run/switches.t |   13 ++++++++++++-
 1 files changed, 12 insertions(+), 1 deletions(-)

diff --git a/t/run/switches.t b/t/run/switches.t
index 5c3b5ee..d946106 100644
--- a/t/run/switches.t
+++ b/t/run/switches.t
@@ -11,7 +11,7 @@ BEGIN {
 
 BEGIN { require "./test.pl"; }
 
-plan(tests => 114);
+plan(tests => 115);
 
 use Config;
 
@@ -107,6 +107,17 @@ SWTEST
     );
 }
 
+{
+    my $tempdir = tempfile;
+    mkdir $tempdir, 0700 or die "Can't mkdir '$tempdir': $!";
+    like(
+        runperl( switches => [ '-c' ], args  => [ $tempdir ], stderr => 1),
+        qr/Can't open perl script.*$tempdir.*is a directory/s,
+        "RT \#61362: Cannot syntax-check a directory"
+    );
+    rmdir $tempdir or die "Can't rmdir '$tempdir': $!";
+}
+
 # Tests for -l
 
 $r = runperl(
-- 
1.6.3.2


```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat Jan 05 17&#8203;:53&#8203;:13 2013\, rjbs wrote&#8203;:
> I have written b1469bd to catch this\.  I haven't written any tests\,
> but can\.  I have tested it by
> hand\.  I thought it worth letting anybody who knows perl\.c better than
> I do \(read&#8203;: anybody\) have
> a look before I spend more time on it\.

$ git diff blead\.\.\.origin/rjbs/cant\-run\-dir |cat
<details><summary>Inline Patch</summary>

```diff
diff --git a/perl.c b/perl.c
index c7e1d54..892ca42 100644
--- a/perl.c
+++ b/perl.c
@@ -3649,6 +3649,7 @@ S_open_script(pTHX_ const char *scriptname, bool

```
</details>
dosearch, bool *suidscript)
&nbsp;     int fdscript = \-1;
&nbsp;     PerlIO \*rsfp = NULL;
&nbsp;     dVAR;
\+    Stat\_t tmpstatbuf;
 
&nbsp;     PERL\_ARGS\_ASSERT\_OPEN\_SCRIPT;
 
@&#8203;@&#8203; \-3758\,6 \+3759\,13 @&#8203;@&#8203; S\_open\_script\(pTHX\_ const char \*scriptname\, bool
dosearch\, bool \*suidscript\)
&nbsp;     /\* ensure close\-on\-exec \*/
&nbsp;     fcntl\(PerlIO\_fileno\(rsfp\)\, F\_SETFD\, 1\);
 \#endif
\+
\+    if \(PerlLIO\_fstat\(PerlIO\_fileno\(rsfp\)\, &tmpstatbuf\) >= 0
\+        && S\_ISDIR\(tmpstatbuf\.st\_mode\)\)
\+        Perl\_croak\(aTHX\_ "Can't open perl script \\"%s\\"&#8203;: %s\\n"\,
\+            CopFILE\(PL\_curcop\)\,
\+            "is a directory"\);
\+
&nbsp;     return rsfp;
 \}

Something about the output is amiss&#8203;:

\#\#\#\#\#
$ mkdir frbulous 

$ \./perl \-c frbulous/
Can't open perl script "frbulous/"&#8203;: is a directory

$ \./perl  frbulous 
Can't open perl script "frbulous"&#8203;: is a directory
\#\#\#\#\#

Wouldn't this be better?

\#\#\#\#\#
$ \./perl  frbulous 
Can't open perl script "frbulous"; it is a directory
\#\#\#\#\#

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @rjbs

\* James E Keenan via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-01\-06T08&#8203;:56&#8203;:34\]
> $ \./perl  frbulous 
> Can't open perl script "frbulous"&#8203;: is a directory
> \#\#\#\#\#
> 
> Wouldn't this be better?
> 
> \#\#\#\#\#
> $ \./perl  frbulous 
> Can't open perl script "frbulous"; it is a directory

The semicolon means we'd need a new perldiag entry\, so no on that front\.  The
missing "it" is probably the result of me reworking that line a pitiable number
of times\.  I'll change it\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @ilmari

Ricardo Signes \<perl\.p5p@&#8203;rjbs\.manxome\.org> writes&#8203;:

> \* James E Keenan via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-01\-06T08&#8203;:56&#8203;:34\]
>> $ \./perl  frbulous 
>> Can't open perl script "frbulous"&#8203;: is a directory
>> \#\#\#\#\#
>> 
>> Wouldn't this be better?
>> 
>> \#\#\#\#\#
>> $ \./perl  frbulous 
>> Can't open perl script "frbulous"; it is a directory
>
> The semicolon means we'd need a new perldiag entry\, so no on that front\.  The
> missing "it" is probably the result of me reworking that line a pitiable number
> of times\.  I'll change it\.

Wouldn'it be best to use the standard strerror\(EISDIR\)\, for consistency?

$ cat /
cat&#8203;: /&#8203;: Is a directory

\-\- 
"A disappointingly low fraction of the human race is\,
 at any given time\, on fire\." \- Stig Sandbeck Mathisen

**Comment by p5pRT** [comment]
### From @rjbs

Fixed in cdd06489088cf64a0e621b79cdf0c48eb55f2622 and previous 3 commits\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Fixed in cdd06489088cf64a0e621b79cdf0c48eb55f2622 and previous 3 commits\.

**Comment by p5pRT** [comment]
@rjbs - Status changed from 'open' to 'resolved'
